File: ros2_workspaces/ros2_from_scratch_by_Renard_ws/src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64

"""
We have a publishers and a timer with a callback.
The timer creates an Int64 message and publishes it to the number topic.

This structure is very common in ROS2. Whenever we want to pubish data from a sensor, 
we use something similar.
"""
class NumberPublisherNode(Node): 

    def __init__(self):
        super().__init__("number_publisher_node") 
        # declare parameters --------------
        self.declare_parameter("number", 2)
        self.declare_parameter("publish_period", 1.0)
        # Get parameters --------------------------
        self.number_ = self.get_parameter("number").value
        self.timer_period_ = self.get_parameter("publish_period").value
        # queue size is the number of messages that can be retained on the topic
        # if the publish rate exceeds the consumption rate
        self.number_publisher_ = self.create_publisher(Int64, "number", 10)
        # Timer fires the callback each second
        self.number_timer_ = self.create_timer(self.timer_period_, self.publish_number)
        self.get_logger().info(self.get_name()+ " started!!")

    def publish_number(self):
        msg = Int64() # construct message instance
        msg.data = self.number_
        self.number_publisher_.publish(msg)

# note that like c++, main belongs to the file, not a class.
def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros2_workspaces/ros2_from_scratch_by_Renard_ws/src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp" // interface for message

class NumberPublisherNode : public rclcpp::Node  {

    public:
    // constructor
    NumberPublisherNode() : Node("number_publisher_node_cpp") {
        RCLCPP_INFO(this->get_logger(), "Creating %s...", this->get_name());
        //number_ = 2;
        this->declare_parameter("number", 2);
        this->declare_parameter("publish_period", 1.0);
        number_ = this->get_parameter("number").as_int();
        double timer_period_ = this->get_parameter("publish_period").as_double();
        number_publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        number_timer_ = this->create_wall_timer(std::chrono::duration<double>(timer_period_), std::bind(&NumberPublisherNode::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "%s created!!", this->get_name());
    }

    private:
    void publishNumber() {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        number_publisher_->publish(msg);
    }
    int number_;
    /**
     * Publisher is a template class. 
     * It requires a message type to indicate the type of data it works with.
     * We use a shared pointer.
     */
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr number_publisher_;
    rclcpp::TimerBase::SharedPtr number_timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}